umbraco/Umbraco-CMS
# V9: Add unique constraints to cmsDictionary and cmsLanguageText tables

## Details

- Adding a unique index on the `key` (_dictionary name_) value on the **cmsDictionary** table;
- Similarly, a unique index constraint is added for the **cmsLanguageText** table on the `languageId` and `UniqueID` ~and `value`~ columns - otherwise, duplicate values can be created for the same dictionary item;
- Silently log the failure of that database migration step, so people don’t have upgrade issues if they had duplicates (incorrect data) - which prevents the indexes from being created;
- If they want the indexes, they first need to remove the duplicates from their database.

### Code changes
- Updating `DictionaryDto` and `LanguageTextDto` to have unique indexes and unique indexes across columns;
- Creating a new V9 migration, where we drop the existing indexes;
- When trying to add the unique constraints, we will first check if there are any duplicates in the data - as the approach for catching an exception when creating the index directly leads to transaction errors: `“This SqlTransaction has completed; It is no longer usable”`;
- Fix tests;

:exclamation: Note: Couldn’t call the shorthand method `AlterColumn<DictionaryDto>(DictionaryDto.TableName, “key”)` or `Alter.Table(DictionaryDto.TableName).AlterColumn(“key”).AsString().Unique().Do()` to re-create the indexes - The column was altered but no indexes were added (even when there were no duplicates).

---

## Test
- Install the backoffice  on `v9/dev` branch;

### Inserting data
- Navigate to the Translation section;
- Create an item under Dictionary (ex: _“Test”_) and add a corresponding value (ex: _“Test item”_);
- Create another one (ex: _“Test1"_ with value _“Another test item”_) - UI prevents you from creating an item with the same name;
- Make a change in the DB - update the record with a duplicate value:
  - Make a selection to check the `pk` value of the _“Test1"_ record and execute the following query:
```sql
  UPDATE [cmsDictionary]
  SET [key] = ‘Test’
  WHERE pk = 2;
```

This will result in incorrect data for the `cmsDictionary` table - 2 entries have the same value in the `key` column.

You can also add items under one of the “Test”s and make them with the same value in the DB.

</br>

Similar thing for `[cmsLanguageText]` table:
- Add a new language in the Settings section;
- Update the values for the new language on the 2 dictionary items - (ex: “Teste”);
- Update the value of `languageId` (check the `pk` of the record you want to update):
```sql
  UPDATE [cmsLanguageText]
  SET [languageId] = 1
  WHERE pk = 3;
```

You will end up having 2 records with the same `languageId` and `UniqueId` ~and `value`~.


### Executing the migration
- Checkout to the PR’s branch;
- Check the existing indexes of **cmsDictionary** table and **cmsLanguageText** table;
  - They should look like this:

> cmsDictionary
<img src="https://user-images.githubusercontent.com/21998037/128899373-f99b7f6c-da0c-47d9-9c72-2714671e6858.png" height="300" align="center">

</br>

> cmsLanguageText
<img src="https://user-images.githubusercontent.com/21998037/128899404-054153c7-82fe-4b1e-8634-70f85a77e15c.png" height="300" align="center">

</br>

- Run the backoffice and start the migration;
- Make sure the migration succeeds and you can continue navigating through the backoffice;
- Check the errors in the Log Viewer (_UPDATE: The 'value' column is no longer part of the unique constraint_):

![err](https://user-images.githubusercontent.com/21998037/128900160-d09e8536-4a0d-4450-9416-9c79d23937f4.PNG)


- Check that the 2 indexes were not added.

</br>

- Go back and delete the duplicate records;
- Run the following script in the database, to reset the migration id back to the current latest as per _v9/dev_:
```sql
update [umbracoKeyValue] set value = '{10F7BB61-C550-426B-830B-7F954F689CDF}'
where [key] = 'Umbraco.Core.Upgrader.State+Umbraco.Core'
```
- Follow the migration steps through the backoffice again;
- This time there should be no errors, as there shouldn't be any duplicates and the indexes should be added:

> cmsDictionary
![cmsDictionary_After](https://user-images.githubusercontent.com/21998037/128900397-943e0200-d715-452b-b581-d4fdd2e3b167.PNG)

</br>

> cmsLanguageText
<img src="https://user-images.githubusercontent.com/21998037/129569010-1d2da48b-2cce-413a-a342-e3b4669b3f03.png" height="280" align="center">
